docsite/rst/developing_program_flow_modules.rst

```diff
@@ -0,0 +1,392 @@
+    new-style python modules you can get the value of the ``syslog_facility``
+    by looking up ``SYSLOG_FACILITY`` in the :envvar:`ANSIBLE_MODULE_CONSTANTS`
+    environment variable.  See the :ref:`ziploader` documentation for details.
+
+.. _ziploader:
+
+ziploader
+^^^^^^^^^
+
+Ziploader differs from :ref:`module_replacer` in that it uses real python
+imports of things in module_utils instead of merely preprocessing the module.
+It does this by constructing a zipfile of the module, files in
+:file:`ansible/module_utils` that are imported by the module, and some
+boilerplate to pass in the constants.  The zipfile is then base64 encoded and
+wrapped in a small python script which unzips the file on the managed node and
+then invokes python on the file.  (We have to wrap the zipfile in the python
+script so that pipelining will work.)
+
+In ziploader, any imports of python modules from the ``ansible.module_utils``
+package trigger inclusion of that python file into the zipfile.  Instances of
+:code:`#<<INCLUDE_ANSIBLE_MODULE_COMMON>>` in the module are turned into
+:code:`from ansible.module_utils.basic import *` and
+:file:`ansible/module-utils/basic.py` is then included in the zipfile.  Files
+that are included from module_utils are themselves scanned for imports of other
+python modules from module_utils to be included in the zipfile as well.
+
+.. warning::
+    At present, there are two caveats to how ziploader determines other files
+    to import.
+
+    * Ziploader cannot determine whether an import should be included if it is
+      a relative import.  Always use an absolute import that has
+      ``ansible.module_utils`` in it to allow ziploader to determine that the
+      file should be included.
+    * Ziploader does not include python packages (directories with
+      :file:`__init__.py`` in them).  Ziploader only works on :file:`*.py`
+      files that are directly in the :file:`ansible/module_utils` directory.
+
+.. _flow_passing_module_args:
+
+Passing args
+~~~~~~~~~~~~
+
+In :ref:`module_replacer`, module arguments are turned into a json-ified
+string and substituted into the combined module file.  In :ref:`ziploader`,
+the json-ified string is placed in the the :envvar:`ANSIBLE_MODULE_ARGS`
+environment variable.  When :code:`ansible.module_utils.basic` is imported
+it places this string in the global variable
+``ansible.module_utils.basic.MODULE_COMPLEX_ARGS`` and removes it from the
+environment.  Modules probably should not access this variable directly.
+Instead, they should instantiate an :class:`AnsibleModule()` and use
+:meth:`AnsibleModule.params` to access the parsed version of the arguments.
+
+.. _flow_passing_module_constants:
+
+Passing constants
+~~~~~~~~~~~~~~~~~
+
+Currently there are three constants passed from the controller to the modules:
+``ANSIBLE_VERSION``, ``SELINUX_SPECIAL_FS`` and ``SYSLOG_FACILITY``.  In
+:ref:`module_replacer`, ``ANSIBLE_VERSION`` and ``SELINUX_SPECIAL_FS`` were
```

Comment:
Changed to:

``ANSIBLE_VERSION``, ``SELINUX_SPECIAL_FS``, and ``SYSLOG_FACILITY``.

``ANSIBLE_VERSION`` and ``SELINUX_SPECIAL_FS`` in the next line is the subject of a new sentence.